packages/ui/VirtualList/VirtualListBase.js

```diff
@@ -23,728 +27,771 @@ const gridListItemSizeShape = PropTypes.shape({
 
-		if (threshold.max > maxPos) {
-			if (maxPos < threshold.base) {
-				threshold.max = threshold.base;
-				threshold.min = -Infinity;
+			if (dataSize <= 0) {
+				moreInfo.firstVisibleIndex = null;
+				moreInfo.lastVisibleIndex = null;
 			} else {
-				threshold.max = maxPos;
-				threshold.min = maxPos - threshold.base;
+				moreInfo.firstVisibleIndex = (Math.floor((primaryPosition - itemSize) / gridSize) + 1) * dimensionToExtent;
+				moreInfo.lastVisibleIndex = Math.min(dataSize - 1, Math.ceil((primaryPosition + clientSize) / gridSize) * dimensionToExtent - 1);
 			}
 		}
-	}
-
-	setScrollPosition (x, y, dirX, dirY) {
-		const
-			{dataSize} = this.props,
-			{firstIndex, numOfItems} = this.state,
-			{isPrimaryDirectionVertical, threshold, dimensionToExtent, maxFirstIndex, scrollBounds} = this,
-			{gridSize} = this.primary,
-			maxPos = isPrimaryDirectionVertical ? scrollBounds.maxTop : scrollBounds.maxLeft,
-			minOfMax = threshold.base,
-			maxOfMin = maxPos - threshold.base;
-		let delta, numOfGridLines, newFirstIndex = firstIndex, pos, dir = 0;
-
-		if (isPrimaryDirectionVertical) {
-			pos = y;
-			dir = dirY;
-		} else {
-			pos = x;
-			dir = dirX;
-		}
 
-		if (dir === 1 && pos > threshold.max) {
-			delta = pos - threshold.max;
-			numOfGridLines = Math.ceil(delta / gridSize); // how many lines should we add
-			threshold.max = Math.min(maxPos, threshold.max + numOfGridLines * gridSize);
-			threshold.min = Math.min(maxOfMin, threshold.max - gridSize);
-			newFirstIndex += numOfGridLines * dimensionToExtent;
-		} else if (dir === -1 && pos < threshold.min) {
-			delta = threshold.min - pos;
-			numOfGridLines = Math.ceil(delta / gridSize);
-			threshold.max = Math.max(minOfMax, threshold.min - (numOfGridLines * gridSize - gridSize));
-			threshold.min = (threshold.max > minOfMax) ? threshold.max - gridSize : -Infinity;
-			newFirstIndex -= numOfGridLines * dimensionToExtent;
+		syncThreshold (maxPos) {
+			const {threshold} = this;
+
+			if (threshold.max > maxPos) {
+				if (maxPos < threshold.base) {
+					threshold.max = threshold.base;
+					threshold.min = -Infinity;
+				} else {
+					threshold.max = maxPos;
+					threshold.min = maxPos - threshold.base;
+				}
+			}
 		}
 
-		if (threshold.min === -Infinity) {
-			newFirstIndex = 0;
-		} else {
-			newFirstIndex = Math.min(maxFirstIndex, newFirstIndex);
-			newFirstIndex = Math.max(0, newFirstIndex);
+		// Native only
+		scrollToPosition (x, y) {
+			this.containerRef.scrollTo(
+				(this.props.rtl && !this.isPrimaryDirectionVertical) ? this.scrollBounds.maxLeft - x : x, y
+			);
 		}
 
-		this.syncThreshold(maxPos);
-		this.scrollPosition = pos;
-		this.updateMoreInfo(dataSize, pos);
-
-		if (firstIndex !== newFirstIndex) {
-			this.setState({firstIndex: newFirstIndex});
-		} else {
-			this.positionItems({updateFrom: firstIndex, updateTo: firstIndex + numOfItems});
+		// JS only
+		setScrollPosition (x, y, dirX, dirY) {
```

Comment:
`dirX`, `dirY` are not used in this function. Please use `...rest`.

Reply:
Well, I think it looks okay as is. :) There are just two arguments, so the current code looks easy to read.